src/main/scala/org/constellation/p2p/Cluster.scala

```diff
@@ -463,18 +461,35 @@ class Cluster[F[_]: Concurrent: Logger: Timer: ContextShift](ipManager: IPManage
       "cluster_leave"
     )
 
-  def broadcastNodeState(): F[Unit] =
+  def broadcastNodeState(nodeState: NodeState): F[Unit] =
     logThread(
-      getNodeState.flatMap { nodeState =>
-        broadcast(_.postNonBlockingUnitF("status", SetNodeStatus(dao.id, nodeState))(C))
-      }.flatTap {
+      broadcast(_.postNonBlockingUnitF("status", SetNodeStatus(dao.id, nodeState))(C)).flatTap {
         _.filter(_._2.isLeft).toList.traverse {
           case (id, e) => Logger[F].warn(s"Unable to propagate status to node ID: $id")
         }
       }.void,
       "cluster_broadcastNodeState"
     )
 
+  def compareAndSet(expected: NodeState, newState: NodeState): F[SetStateResult] =
+    compareAndSet(Set(expected), newState)
+
+  def compareAndSet(expected: Set[NodeState], newState: NodeState): F[SetStateResult] =
+    nodeState.modify { current =>
+      if (expected.contains(current)) (newState, SetStateResult(current, isNewSet = true))
+      else (current, SetStateResult(current, isNewSet = false))
+    }.flatTap(
+        res =>
+          if (res.isNewSet) LiftIO[F].liftIO(dao.metrics.updateMetricAsync[IO]("nodeState", newState.toString))
+          else Sync[F].unit
+      )
+      .flatTap(
+        res =>
+          if (res.isNewSet && NodeState.broadcastStates.contains(newState) && NodeState.broadcastStates
+                .contains(res.oldState)) broadcastNodeState(newState)
```

Comment:
it means that both old and new state must be broadcastable? shouldn't it be only new one?

Reply:
example: 
pendingDownload -> ready (brodcasted)
ready -> snapshotCreation (not needed to broadcast)
snapshotCreation -> ready (not needed to broadcast)

however if we need to use different state 
pendingDownload -> ready (brodcasted)
ready -> snapshotCreation (not needed to broadcast)
snapshotCreation -> leaving (this shouldbe broadcasted anyway :thinking: )

Reply:
this one need a fix

Reply:
fixed